sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -253,6 +257,143 @@ case class FilterExec(condition: Expression, child: SparkPlan)
+
+  private val physicalRecursiveTerms = new LinkedBlockingQueue[SparkPlan]
+
+  def recursiveTermIterations: Seq[SparkPlan] =
+    physicalRecursiveTerms.toArray(Array.empty[SparkPlan])
+
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+
+  /**
+   * Notify the listeners of the physical plan change.
+   */
+  private def onUpdatePlan(executionId: Long): Unit = {
+    val queryExecution = SQLExecution.getQueryExecution(executionId)
+    sparkContext.listenerBus.post(SparkListenerSQLAdaptiveExecutionUpdate(
+      executionId,
+      queryExecution.toString,
+      SparkPlanInfo.fromSparkPlan(queryExecution.executedPlan)))
+  }
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    val executionIdLong = Option(executionId).map(_.toLong)
+
+    val levelLimit = conf.getConf(SQLConf.CTE_RECURSION_LEVEL_LIMIT)
+
+    // TODO: cache before count, as the RDD can be reused in the next iteration
+    var prevIterationRDD = anchorTerm.execute().map(_.copy())
+    var prevIterationCount = prevIterationRDD.count()
+
+    val accumulatedRDDs = mutable.ArrayBuffer.empty[RDD[InternalRow]]
+    var accumulatedCount = 0L
+
+    val numOutputRows = longMetric("numOutputRows")
+
+    var level = 0
+    while (prevIterationCount > 0) {
+      if (levelLimit != -1 && level > levelLimit) {
+        throw new SparkException(s"Recursion level limit ${levelLimit} reached but query has not " +
+          s"exhausted, try increasing ${SQLConf.CTE_RECURSION_LEVEL_LIMIT.key}")
+      }
+
+      accumulatedRDDs += prevIterationRDD
+      accumulatedCount += prevIterationCount
+
+      numOutputRows += prevIterationCount
+      SQLMetrics.postDriverMetricUpdates(sparkContext, executionId, metrics.values.toSeq)
+
+      val newLogicalRecursiveTerm = logicalRecursiveTerm.transform {
+        case rr @ RecursiveReference(name, _, accumulated, _, _, _) if name == cteName =>
+          val (newStatistics, newRDD) = if (accumulated) {
+            (
+              Statistics(
+                EstimationUtils.getSizePerRow(output) * accumulatedCount,
+                Some(accumulatedCount)
+              ),
+              if (accumulatedRDDs.size > 1) {
+                sparkContext.union(accumulatedRDDs)
+              } else {
+                accumulatedRDDs.head
+              }
+            )
+          } else {
+            (
+              Statistics(
+                EstimationUtils.getSizePerRow(output) * prevIterationCount,
+                Some(prevIterationCount)
+              ),
+              prevIterationRDD
+            )
+          }
+          rr.withNewIteration(level, newStatistics, newRDD)
+      }
+
+      val physicalRecursiveTerm =
+        QueryExecution.prepareExecutedPlan(sqlContext.sparkSession, newLogicalRecursiveTerm)
+
+      physicalRecursiveTerms.offer(physicalRecursiveTerm)
```

Comment:
what is this for?

Reply:
No, you are right, this is not need to be blocking and actually this is not even accessed concurrently so a simple `ListBuffer` would be good.

Reply:
Done.